### PR TITLE
New logger: configuration file loader

### PR DESCRIPTION
**Pull request by MBoldyrev**
### Description of the Change

This change adds the new logger configuration tree structure loading. Also the configuration file parsing is refactored.

### Benefits

One more step towards the new logger. Also the config file reading is generalized with hierarchy support, so the whole JSON object may now be a section of a bigger JSON config. The exact place of an error in the config hierarchy is reported.

Also the `rapidjson` library is removed from the config loader header file.

### CODE REVIEW

File: irohad/main/CMakeLists.txt [hunk 1]
```diff
@@ -62,6 +62,13 @@ target_link_libraries(irohad
     rapidjson
     keys_manager
     common
+    iroha_conf_loader
     )
 
+add_library(iroha_conf_loader iroha_conf_loader.cpp)
```
Comment (Akvinikym) [inline irohad/main/CMakeLists.txt]:
```suggestion
add_library(iroha_conf_loader 
  iroha_conf_loader.cpp)
```
This formatting should be in other changes as well

File: irohad/main/application.hpp [hunk 2]
```diff
@@ -65,6 +65,7 @@ class Irohad {
    * @param proposal_delay - maximum waiting time util emitting new proposal
    * @param vote_delay - waiting time before sending vote to next peer
    * @param keypair - public and private keys for crypto signer
+   * @param logger - the logger to use
```
Comment (Akvinikym) [inline irohad/main/application.hpp]:
```suggestion
   * @param logger_manager - the logger to use
```

File: irohad/main/iroha_conf_loader.cpp [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+              "The place of error is out of the printed string boundaries!");
+
+/**
+ * Adds the children logger configs from parent logger JSON object to parent
+ * logger config. The parent logger JSON object is searched for the children
+ * config section, and the children configs are parsed and created if the
+ * section is present.
+ * @param parent_config - the parent logger config
+ * @param parent_obj - the parent logger json configuration
+ */
```
Comment (Akvinikym) [inline irohad/main/iroha_conf_loader.cpp]: Please, update all the docs in this file, a lot of them have wrong arguments

File: irohad/main/iroha_conf_loader.cpp [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+/**
+ * Gets a value by a key from a JSON object, if present.
+ * @param dest - the variable to store the value
+ * @param obj - the source JSON object
+ * @param key - the key for the requested value
+ * @return true if the value was loaded, otherwise false.
+ */
+template <typename TDest, typename TKey>
+bool tryGetValByKey(const std::string &path,
+                    TDest &dest,
```
Comment (Akvinikym) [inline irohad/main/iroha_conf_loader.cpp]: I think, template parameters here and in other functions in this file should be accepted by a universal reference, not a lvalue one

File: irohad/main/CMakeLists.txt (hunk 1 above)
Comment (MBoldyrev) [reply irohad/main/CMakeLists.txt]: I see that throughout the project the 1-line syntax is used in such cases.

File: irohad/main/iroha_conf_loader.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+#include <rapidjson/error/en.h>
+#include <rapidjson/istreamwrapper.h>
+#include <rapidjson/rapidjson.h>
+#include <boost/algorithm/string/join.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include "main/iroha_conf_literals.hpp"
+
+#include "main/assert_config.hpp"
+
+namespace ac = assert_config;
```
[the remark this answers is not in the document]
Comment (MBoldyrev) [reply irohad/main/iroha_conf_loader.cpp]: Removed them both and the `irohad/main/irohad.cpp` file.

**Pull request merged**